API end points That pass Test

LarryKarani opened the pull request:
**What does this PR do?**
Refactors the code to pass the test by adding API end points

**Endpoints added**

EndPoint | Functionality
-- | --
POST /api/v1/rides | Creates a new ride
GET /api/v1/rides | Fetches all available rides
GET /api/v1/rides/<rideId> | Fetches a single ride
POST /api/v1/rides/<rideId>/requests | Creates a ride offer
POST/api/v1/register| Registers a new user
GET  /api/v1/users| Gets all the registered users
POST/api/v1/login| logs in a registered user

**Task achieved in this PR**
- create API end to get all available ride offers
- create API end point to get a specific ride offer
- create API end point to create a ride offer
- create API  to request to join a ride offer
- create API end point to login and authenticate a user
- create API end point to add a new user
- create API end point to get all users

**How should this be manually tested?**

Clone this repo using git clone 
Check the ReadMe for instructions on setting up the project.
git checkout f1
test the above endpoints using post-man

**Relevant Pivotal stories**
https://www.pivotaltracker.com/n/projects/2180294

![ure](https://user-images.githubusercontent.com/22986772/41918043-3116256c-7964-11e8-8bda-b02289038c0d.PNG)

LarryKarani committed:
add test for create ride offer ap1, request ride api and the ride offer
 models with regards to TDD concepts finishes[#158552546]
 finishes[#158527674] finishes[#158528050]

LarryKarani committed:
Refactored the code to pass the writen test
 -add utils file that contains validatos which validate user input data
 -add API end points
 -add jwt authetification to secure endpoints
 finishes[#158527407] [#15828247] [#158527914] [#158528596]

LarryKarani committed:
removed spaces empty spaces and comment code

JoshuaOndieki commented on FlaskApp/api/rideoffers.py:
```diff
@@ -57,7 +55,4 @@ def get_specific_offer(id):
-    
-
-
+                return{"message": " rideoffer does not exit"}
```
Add a newline at the end of file

JoshuaOndieki commented on FlaskApp/api/utils.py:
```diff
@@ -0,0 +1,70 @@
+    def validate(self, obj, activity):
+        if activity == 'create_ride':
+            for prop in self.ride_ofer_props:
+                if prop not in obj:
```
You could split your validators into different functions instead of using an if-else statement

LarryKarani replied:
Noted

JoshuaOndieki commented on FlaskApp/api/app.py:
```diff
@@ -0,0 +1,24 @@
+    api.add_resource(AllRides, '/api/v1/rides')
+    api.add_resource(GetRide, '/api/v1/rides/<rideId>')
+    api.add_resource( JoinRequest, '/api/v1/rides/<rideId>/requests')
+    api.add_resource(Usersg, '/api/v1/users')
```
When the app scales, you'll have a lot of URLs. I suggest you register your URLs in a blueprint and only register the blueprints in your create_app

LarryKarani replied:
Thanks ill work on that

JoshuaOndieki reviewed (changes requested):
Getting a plugin or an editor that will automatically add newlines at the end of your files would make your work easier

LarryKarani committed:
add travis ci

LarryKarani committed:
added build status on read me

LarryKarani committed:
added authetication test

LarryKarani committed:
edited yml to capture brances

justmesam commented on FlaskApp/api/user.py:
```diff
@@ -0,0 +1,30 @@
+from datetime import datetime 
+class User:
+    #class variable holds all users
+    users = []
```
You have a class user for users I guess? but I cant see where your `__init__` is  it makes me wonder how are you going to instantiate the user when you get the values of the user.

justmesam commented on FlaskApp/api/user.py:
```diff
@@ -0,0 +1,30 @@
+    #class variable holds all users
+    users = []
+
+    def regester_user(self, name, username, email, password):
```
check for a typo on your method name

justmesam commented on FlaskApp/tests/test_rideoffers.py:
```diff
@@ -0,0 +1,46 @@
+import unittest
+from FlaskApp.api.rideoffers import RideOffer
+
+class RideOfferTestCase(unittest.TestCase):
```
I don't think these tests are of relevance since you are already testing the endpoints so this should be covered. You are just repeating yourself.

justmesam commented on FlaskApp/api/user.py:
```diff
@@ -0,0 +1,30 @@
+from datetime import datetime 
+class User:
+    #class variable holds all users
+    users = []
```
You have a class user for users I guess? but I cant see where your `__init__` is it makes me wonder how are you going to instantiate the user when you get the values of the user.

LarryKarani replied:
Okey this is how i hundled the logic. The create user method returns a dictionary. Then directly appends the  user dictionary to the users list hence the list contains a list of dictionary containing user info.

justmesam replied:
So with that logic how are you implementing the basics of `OOP`??

LarryKarani replied:
Noted ill use an __init__ to instantiate the user attributes Thanks for the feedback

justmesam commented on FlaskApp/api/routes.py:
```diff
@@ -0,0 +1,198 @@
+        return response
+
+class JoinRequest(Resource):
+    @jwt_required
```
you implemented `jwt`?

LarryKarani replied:
Yes I did implement jwt

LarryKarani committed:
chaged yml

justmesam commented on FlaskApp/api/request.py:
```diff
@@ -0,0 +1,35 @@
+rideoffer = RideOffer()
+
+
+class RequestsJ:
```
How does this class follow the `OOP` concept??

justmesam commented on FlaskApp/api/routes.py:
```diff
@@ -0,0 +1,198 @@
+validator = Validator()
+join_req =  RequestsJ()
+
+user_r = User()
```
This is not how `OOP` is used to create instances of `RideOffer(), Validator(), RequestsJ(), User()` we use it to create many instances of whatever we are trying to achieve but not a single instance used all over.

justmesam reviewed (changes requested):
check on the issues raised

LarryKarani committed:
Implements OOP concepts in models:
   -add __init__() method to initialize the class
   -removed unnesesary instance of classes
   -formated test to work with OOP model

LarryKarani committed:
fixed a typo

LarryKarani committed:
edited travis
